azurerm/resource_arm_application_gateway.go

```diff
@@ -0,0 +1,1685 @@
+	d.Set("backend_http_settings", flattenApplicationGatewayBackendHTTPSettings(ApplicationGateway.ApplicationGatewayPropertiesFormat.BackendHTTPSettingsCollection))
+	d.Set("http_listener", flattenApplicationGatewayHTTPListeners(ApplicationGateway.ApplicationGatewayPropertiesFormat.HTTPListeners))
+	d.Set("probe", flattenApplicationGatewayProbes(ApplicationGateway.ApplicationGatewayPropertiesFormat.Probes))
+	d.Set("request_routing_rule", flattenApplicationGatewayRequestRoutingRules(ApplicationGateway.ApplicationGatewayPropertiesFormat.RequestRoutingRules))
+	d.Set("url_path_map", flattenApplicationGatewayURLPathMaps(ApplicationGateway.ApplicationGatewayPropertiesFormat.URLPathMaps))
+	d.Set("authentication_certificate", schema.NewSet(hashApplicationGatewayAuthenticationCertificates, flattenApplicationGatewayAuthenticationCertificates(ApplicationGateway.ApplicationGatewayPropertiesFormat.AuthenticationCertificates)))
+	d.Set("ssl_certificate", schema.NewSet(hashApplicationGatewaySslCertificates, flattenApplicationGatewaySslCertificates(ApplicationGateway.ApplicationGatewayPropertiesFormat.SslCertificates)))
+
+	if ApplicationGateway.ApplicationGatewayPropertiesFormat.WebApplicationFirewallConfiguration != nil {
+		d.Set("waf_configuration", schema.NewSet(hashApplicationGatewayWafConfig,
+			flattenApplicationGatewayWafConfig(ApplicationGateway.ApplicationGatewayPropertiesFormat.WebApplicationFirewallConfiguration)))
+	}
+
+	flattenAndSetTags(d, ApplicationGateway.Tags)
+
+	return nil
+}
+
+func resourceArmApplicationGatewayDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).applicationGatewayClient
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return errwrap.Wrapf("Error Parsing Azure Resource ID {{err}}", err)
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["applicationGateways"]
+
+	_, errChan := client.Delete(resGroup, name, make(chan struct{}))
+	err = <-errChan
+	if err != nil {
+		return errwrap.Wrapf("Error Deleting ApplicationGateway {{err}}", err)
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func ApplicationGatewayResGroupAndNameFromID(ApplicationGatewayID string) (string, string, error) {
+	id, err := parseAzureResourceID(ApplicationGatewayID)
+	if err != nil {
+		return "", "", err
+	}
+	name := id.Path["applicationGateways"]
+	resGroup := id.ResourceGroup
+
+	return resGroup, name, nil
+}
+
+func retrieveApplicationGatewayById(ApplicationGatewayID string, meta interface{}) (*network.ApplicationGateway, bool, error) {
+	client := meta.(*ArmClient).applicationGatewayClient
+
+	resGroup, name, err := ApplicationGatewayResGroupAndNameFromID(ApplicationGatewayID)
+	if err != nil {
+		return nil, false, errwrap.Wrapf("Error Getting ApplicationGateway Name and Group: {{err}}", err)
+	}
+
+	resp, err := client.Get(resGroup, name)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
```

Comment:
there's a new helper method for this which checks for `nil` values, can we switch this to using:

```
if responseWasNotFound(resp.Response) {
 ...
}
```